### PR TITLE
Pre-training on masked patches restoring

### PR DESCRIPTION
**Pull request by dkadochnikov**
PIXEL-like pre-train on masked patches

### CODE REVIEW

File: src/main.py [hunk 1]
```diff
@@ -122,13 +129,22 @@ def train_vtr_encoder(args: Namespace, train_data: list, val_data: list = None,
             f"# classes: {len(char2array.keys())}"
         )
         ocr = OCRHead(
-            input_size=vtr.out_channels * (vtr.font_size // vtr.pool_kernel_size ** (len(channels) - 1)),
+            input_size=16,
```
Comment (borisshapa) [inline src/main.py:132]: Why we hard code this number?

File: src/utils/masking.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
+import random
+from typing import List, Optional, Union
+
+import numpy as np
+
+
+class SpanMaskingGenerator:
+    """
```
Comment (borisshapa) [inline src/utils/masking.py:8]: It's really good that you write docstrings!

File: src/models/pretraining.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        n_head: int = 8,
+        n_layers: int = 4,
+        device: str = "cpu",
+        ocr: OCRHead = None,
+        char2int_dict: dict = None,
+        alpha: float = 1,
+    ):
+        super().__init__()
+        config = BertConfig(hidden_size=emb_size, num_attention_heads=n_head)
+        self.encoder = BertModel(config)
```
Comment (borisshapa) [inline src/models/pretraining.py:27]: I expected that we would use [TransformerEnocder](https://pytorch.org/docs/stable/generated/torch.nn.TransformerEncoder.html) instead of adding a dependency from HF

File: src/models/pretraining.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+class Pretrain(nn.Module):
+    def __init__(
+        self,
+        emb_size: int = 512,
+        n_head: int = 8,
+        n_layers: int = 4,
+        device: str = "cpu",
+        ocr: OCRHead = None,
+        char2int_dict: dict = None,
```
Comment (borisshapa) [inline src/models/pretraining.py:22]: It isn't a good practice to add the type of the variable to its name: 
https://google.github.io/styleguide/pyguide.html#3161-names-to-avoid (the last point)

You've already had this information in the typing

File: src/models/pretraining.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        self.emb_size = emb_size
+        self.alpha = alpha
+        self.iter = 0
+
+    def forward(self, input_batch: dict[str, list | torch.Tensor]):
+
+        batch_size, slice_count, height, width = input_batch["slices"].shape
+        slices = input_batch["slices"].view(batch_size, slice_count, height * width)
+        pos_slices = self.positional_enc(slices)
+        slices_detached = pos_slices.detach()
```
Comment (borisshapa) [inline src/models/pretraining.py:50]: Why is detaching required here?

File: src/models/pretraining.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+        grey_slice = torch.full((height, width), 128, dtype=torch.float32, device=self.device).flatten()
```
Comment (borisshapa) [inline src/models/pretraining.py:52]: Let's take 128 as a constant.

File: src/utils/masking.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        self.cumulative_span_weights = cumulative_span_weights
+
+    def _mask(self, mask, max_mask_patches):
+        delta = 0
+        # Lower number of attempts will speed up mask generation but might cause a lot fewer patches to be masked
+        # than desired, particularly for high masking ratios
+        for attempt in range(100):
+            # Randomly select span length within specified range
+            span = random.choices(self.span_range, cum_weights=self.cumulative_span_weights, k=1)[0]
+            if span < self.num_patches:
```
Comment (borisshapa) [inline src/utils/masking.py:58]: Am I right that if we chose the span size from the range(1, max_mask_patches + 1) one line above, than we don't need the check here?

File: src/models/pretraining.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        masks = []
+        unmasked_slices = []
+        unmasked_texts = []
+        for i in range(batch_size):
+            mask = self.masking(len(slices[i]))
+            masks.append(mask)
+            masked_idx = np.where(mask == 1)[0]
+            masked_idx = masked_idx[masked_idx < len(slices_detached[i])]
+            if masked_idx.size < 10:
```
Comment (borisshapa) [inline src/models/pretraining.py:61]: 10?

File: src/models/pretraining.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+        encoded_text = self.positional_dec(encoded_text[0])
+        decoded_text = self.decoder(encoded_text)
+
+        masked_slices = []
+        masked_originals = []
+        for i in range(batch_size):
+            masked_idx = np.where(masks[i] == 1)[0]
+            masked_idx = masked_idx[masked_idx < len(slices_detached[i])]
+            if masked_idx.size < 10:
+                masked_idx = np.arange(10)
```
Comment (borisshapa) [inline src/models/pretraining.py:83]: I think that it's worth to save the masked_idx to a variable and don't copy-paste these 3 lines

File: src/models/pretraining.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+        result = {"loss": self.criterion(masked_slices, masked_originals).sum()}
+        if self.ocr:
+            unmasked_slices = torch.stack(unmasked_slices)
+            slice_count = unmasked_slices.shape[1]
+            unmasked_slices = unmasked_slices.view(batch_size * slice_count, 1, height, width)
+            result["ctc_loss"] = compute_ctc_loss(
+                self.ctc_criterion, self.ocr, unmasked_slices, unmasked_texts, self.char2int_dict
+            )
+            result["lpips_loss"] = result["loss"]
+            result["loss"] += self.alpha * result["ctc_loss"]
```
Comment (borisshapa) [inline src/models/pretraining.py:102]:
```
        result = {"lpips_loss": self.criterion(masked_slices, masked_originals).sum()}
         if self.ocr:
             unmasked_slices = torch.stack(unmasked_slices)
             slice_count = unmasked_slices.shape[1]
             unmasked_slices = unmasked_slices.view(batch_size * slice_count, 1, height, width)
             result["ctc_loss"] = compute_ctc_loss(
                 self.ctc_criterion, self.ocr, unmasked_slices, unmasked_texts, self.char2int_dict
             )
             result["loss"] = result["lpips_loss"] + self.alpha * result["ctc_loss"]
```

File: src/utils/common.py [hunk 11]
```diff
@@ -128,14 +128,49 @@ def compute_ctc_loss(
...
+        if not cur_masked.intersection(neighbours_idx) and not cur_neighbours.intersection(masked_patches_idx):
+            masked_patches_idx.update(cur_masked)
+            neighbours_idx.update(cur_neighbours)
+            failures = 0
+        else:
+            failures += 1
+
+    mask[list(masked_patches_idx)] = 1
+
+    return mask
```
Comment (SpirinEgor) [inline src/utils/common.py:171]: That is very bad and slow way to create mask :(

First of all, you need to think in matricies. Let this function take batch size and seq len as arguments.
Secondly, max span is an important parameter, it must also be passed as an argument.

```python
def create_noise_mask(batch_size, seq_len, noise_density=0.2, span_lens=6):
    # 1. Create random mask
    rand_nums = torch.rand(batch_size, seq_len)
    mask = rand_nums.le(noise_density)

    # 2. Calculate position of each mask in span
    cum_sum = torch.cumsum(mask, dim=-1)
    cum_sum_not_mask = torch.masked_fill(cum_sum, mask, 0)
    mask_position = cum_sum - torch.cummax(cum_sum_not_mask, dim=-1).values

    # 3. Avoid too long spans
    mask[mask_position > span_lens] = False
    return mask
```

Please, double check my code. It is good to write here stress tests, e.g. generate 1M mask and natively check all conditions.

File: src/utils/common.py [hunk 12]
```diff
@@ -154,7 +189,7 @@ class PositionalEncoding(nn.Module):
     def __init__(self, hidden_size: int, dropout=0.1, max_len=512):
         super(PositionalEncoding, self).__init__()
 
-        self.dropout = nn.Dropout(p=dropout)
+        #self.dropout = nn.Dropout(p=dropout)
```
Comment (SpirinEgor) [inline src/utils/common.py:192]: Why do you remove dropout here?

File: src/utils/masking.py [hunk 13]
```diff
@@ -0,0 +1,115 @@
+import random
+from typing import List, Optional, Union
+
+import numpy as np
+
+
+class SpanMaskingGenerator:
```
Comment (SpirinEgor) [inline src/utils/masking.py:7]: It seems that this class is unused. If yes, please remove it

File: src/models/pretraining.py [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+        grey_slice = torch.full((height, width), 128, dtype=torch.float32, device=self.device).flatten()
+        masks = []
+        unmasked_slices = []
+        unmasked_texts = []
+        for i in range(batch_size):
+            not_padded_len = torch.sum(input_batch["attention_mask"][i], dtype=torch.int).item()
+            mask = masking(not_padded_len)
+            masks.append(mask)
+            masked_idx = np.where(mask == 1)[0]
+            slices[i][masked_idx] = grey_slice
```
Comment (SpirinEgor) [inline src/models/pretraining.py:59]: See comments in `common.py` file, I provide effective implementation to create mask for entire batch.
So:

```python
# [batch size; slice count], `True` for patches to mask 
mask = create_noise_mask(batch_size, slice_count, self.noise_density, self.span_lens)
mask *= input_batch["attention_mask"]  # attn mask has the same shape and `True` for non padding values

slices[mask, :, :] = self.MASK_VALUE  # this would fill all masked tokens with specified value, e.g 128 
```

**Review by SpirinEgor — CHANGES_REQUESTED** [review]

File: src/utils/common.py [hunk 15]
```diff
@@ -118,24 +117,89 @@ def _set_seed(seed: int):
...
-    criterion: torch.nn.modules.loss.CTCLoss, ocr: OCRHead, embeddings: torch.Tensor, texts: list, char2int_dict: dict
+    criterion: torch.nn.modules.loss.CTCLoss, ocr: OCRHead, embeddings: torch.Tensor, texts: list, char2int: dict
 ):
     logits = ocr(embeddings)
     log_probs = torch.nn.functional.log_softmax(logits, dim=2)
     input_lengths = torch.LongTensor([log_probs.shape[0]] * log_probs.shape[1])
 
     chars = list("".join(np.concatenate(texts).flatten()))
-    targets = torch.LongTensor([char2int_dict.get(c, char2int_dict["UNK"]) for c in chars])
+    targets = torch.LongTensor([char2int[c] for c in chars])
```
Comment (SpirinEgor) [inline src/utils/common.py:127]: why do you remove safe get with UNK token?

File: src/utils/common.py [hunk 16]
```diff
@@ -118,24 +117,89 @@ def _set_seed(seed: int):
...
+    return mask
+
+
+def plot_slices(
+    decoded: tuple[torch.Tensor, torch.Tensor],
+    orig: tuple[torch.Tensor, torch.Tensor],
+    iter_num: int,
+    loss: float,
+    folder_name: str = None,
+):
```
Comment (SpirinEgor) [inline src/utils/common.py:184]: Btw, since we are using wandb to log train polts, we can also log images
https://docs.wandb.ai/guides/track/log/media

**Review by SpirinEgor — APPROVED** [review]
